core/src/main/scala/kafka/utils/LoggingController.scala

```diff
@@ -27,22 +28,55 @@ import java.util.Locale
 import scala.jdk.CollectionConverters._
 
 
-object Log4jController {
+object LoggingController {
 
   /**
-   * Note: In log4j, the root logger's name was "root" and Kafka also followed that name for dynamic logging control feature.
+   * Note: In Log4j 1, the root logger's name was "root" and Kafka also followed that name for dynamic logging control feature.
    *
    * The root logger's name is changed in log4j2 to empty string (see: [[LogManager.ROOT_LOGGER_NAME]]) but for backward-
    * compatibility. Kafka keeps its original root logger name. It is why here is a dedicated definition for the root logger name.
    */
   val ROOT_LOGGER = "root"
 
+  private[this] val delegate: LoggingControllerDelegate = {
+    try {
+      new Log4jCoreController
+    } catch {
+      case _: ClassCastException | _: LinkageError => new NoOpController
+    }
+  }
+
   /**
    * Returns a map of the log4j loggers and their assigned log level.
    * If a logger does not have a log level assigned, we return the log level of the first ancestor with a level configured.
    */
-  def loggers: Map[String, String] = {
-    val logContext = LogManager.getContext(false).asInstanceOf[LoggerContext]
+  def loggers: Map[String, String] = delegate.loggers
+
+  /**
+   * Sets the log level of a particular logger. If the given logLevel is not an available level
+   * (i.e., one of OFF, FATAL, ERROR, WARN, INFO, DEBUG, TRACE, ALL) it falls back to DEBUG.
+   *
+   * @see [[Level.toLevel]]
+   */
+  def logLevel(loggerName: String, logLevel: String): Boolean = delegate.logLevel(loggerName, logLevel)
+
+  def unsetLogLevel(loggerName: String): Boolean = delegate.unsetLogLevel(loggerName)
+
+  def loggerExists(loggerName: String): Boolean = delegate.loggerExists(loggerName)
+}
+
+private class NoOpController extends LoggingControllerDelegate {
+  override def loggers: Map[String, String] = Map.empty
+
+  override def logLevel(loggerName: String, logLevel: String): Boolean = false
+
+  override def unsetLogLevel(loggerName: String): Boolean = false
+}
+
+private class Log4jCoreController extends LoggingControllerDelegate {
```

Comment:
ditto. maybe we can move it to another class file to ensure log4j code is used by the class file only.

Reply:
Fixed in https://github.com/apache/kafka/pull/18496/commits/db05e49b49836e9392aaf1f3c0d6a27124ec4445